MDEV-30188: Ensure all binlog* variables are visible as system variables

DarrenZhang01 opened the pull request:
*The Jira issue number for this PR is: MDEV-30188

## Description
Turn the remaining three `binlog*` options binlog_do_db, binlog_ignore_db,
binlog_rows_event_max_size into global variables so that they can be
visible from the SQL user level. This is for audit / secure
configuration check purposes.

Create new MTR tests to make sure that the newly created global
variables can be visible from the command line interface.

Behavior before the code change:

    MariaDB [(none)]> SHOW GLOBAL VARIABLES WHERE
        -> Variable_name LIKE 'binlog_do_db' OR
        -> Variable_name LIKE 'binlog_ignore_db' OR
        -> Variable_name LIKE 'binlog_row_event_max_size';
    Empty set (0.001 sec)

Behavior after the code change:

    MariaDB [(none)]> SHOW GLOBAL VARIABLES WHERE
        -> Variable_name LIKE 'binlog_do_db' OR
        -> Variable_name LIKE 'binlog_ignore_db' OR
        -> Variable_name LIKE 'binlog_row_event_max_size';
    +---------------------------+-------+
    | Variable_name             | Value |
    +---------------------------+-------+
    | binlog_do_db              |       |
    | binlog_ignore_db          |       |
    | binlog_row_event_max_size | 8192  |
    +---------------------------+-------+
    3 rows in set (0.001 sec)


## How can this PR be tested?

This PR can be tested using the newly created MTR test files `binlog_do_db.test`, `binlog_ignore_db.test` and `binlog_rows_event_max_size.test`. All of the aforementioned tests have passed.

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*

## Copyright
All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2259) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2259) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2259) it.</sub>

dlenski commented on sql/sys_vars.cc:
```diff
@@ -3088,6 +3088,24 @@ static Sys_var_ulong Sys_query_prealloc_size(
+     "grouped into events smaller than this size if possible. "
+     "The value has to be a multiple of 256.",
+     GLOBAL_VAR(opt_binlog_rows_event_max_size), CMD_LINE(REQUIRED_ARG),
+     VALID_RANGE(256, UINT_MAX32 - (UINT_MAX32 % 256)), DEFAULT(8192),
```
Note that this differs from the previously-implemented maximum:

- Previous maximum: `UINT_MAX32 - 1` = `0xFFFFFFFF - 1` = `0xFFFFFFFE` (not a multiple of 256 :man_shrugging:)
- New maximum: `UINT_MAX32 - (UINT_MAX32 % 256)` = `0xFFFFFFFF - (0xFFFFFFFF % 0xFF)` = `0xFFFFFF00` (now a multiple of 256)

PhysicsTing commented on mysql-test/main/mysqld--help.result:
```diff
@@ -1503,9 +1503,11 @@ binlog-checksum CRC32
 binlog-expire-logs-seconds 0
 binlog-file-cache-size 16384
 binlog-format MIXED
+binlog-ignore-db (No default value)
```
Good job on fixing the help output! Now we not only support these two as variables, but also maintain good backward compatibility (ie. "--binlog-do-db-db1 --binlog-do-db=db2" from command line still works)!

ottok commented:
@LinuxJedi Could you please review this? It has been 28 days without any upstream response

bnestere commented on mysql-test/suite/sys_vars/t/binlog_ignore_db.test:
```diff
@@ -0,0 +1,15 @@
+# Restart the server with "--binlog_ignore_db='1'" option
+--let $restart_parameters = "--binlog_ignore_db='1'"
+--source include/restart_mysqld.inc
+SELECT @@binlog_ignore_db AS EXPECT_1;
```
For `binlog_do_db` and `binlog_ignore_db` tests, can you add another test case to ensure multiple databases are composed into a single comma separated list correctly?

DarrenZhang01 replied:
Thanks @bnestere ! That's a good idea! (Actually I tested this case manually before but probably should also add it into the test file)

bnestere commented on sql/mysqld.cc:
```diff
@@ -8723,7 +8783,6 @@ static int get_options(int *argc_ptr, char ***argv_ptr)
       double2ulonglong(slave_max_statement_time_double * 1e6);
   }
 #endif
-
```
Can you add this line back in to prevent potential merge problems?

bnestere reviewed (changes requested):
Hi @DarrenZhang01 !

Thanks for the contribution! I left a few notes.

bnestere commented on sql/mysqld.cc:
```diff
@@ -5785,6 +5794,63 @@ int mysqld_main(int argc, char **argv)
+  // Assign the concatenated string to the global variable `binlog_ignore_db`.
+  binlog_ignore_db = new char[binlog_ignore_db_string_concatenation.size() + 1];
+  // unsigned LENGTH_BINLOG_IGNORE_DB = binlog_ignore_db_string_concatenation.size() + 1;
+  strcpy(binlog_ignore_db, binlog_ignore_db_string_concatenation.c_str());
```
Sure!

The replication filter variables only exist in the global scope, and thereby cannot be updated in the session scope. Translating this to implementation, `Sys_var_rpl_filter` implements the `global_update()` function; however, `session_update()` simply returns in error:
```
  bool session_update(THD *thd, set_var *var)
  {
    DBUG_ASSERT(FALSE);
    return true;
  }
```

In our version, `Sys_var_binlog_filter`, we don't want to allow global updates, and therefore its implementation of `global_update()` should behave the same way, i.e.
```
  bool global_update(THD *thd, set_var *var)
  {
    DBUG_ASSERT(FALSE);
    return false;
  }
```

And the same for our version of the function `global_save_default`, with its corresponding implementation being `session_save_default`.

Does that make more sense?

DarrenZhang01 replied:
Thanks @bnestere ! I will think about the new structure and get back to you with updates in about 3 weeks!

DarrenZhang01 replied:
Hi @bnestere , quick question. Do you think that the `set_filter_value` and `global_value_ptr` function need to be updated in `Sys_var_binlog_filter`? These two functions seem to be specifically designed for their own set of variables. Thanks!

bnestere commented:
Hi @DarrenZhang01!

I have a question for you. With the upcoming release of 10.11, we will be moving this task to target 10.12. Do you have the bandwidth to extend this PR to cover the additional replication variables mentioned in MDEV-28672? Specifically, the following list below (note that `replicate-rewrite-db` was removed from the list as it was added in MDEV-15530). Also note that the 10.12 deadline is December 13.
```
master-info-file
master-retry-count
replicate-same-server-id
show-slave-auth-info
skip-slave-start
```

DarrenZhang01 commented:
Hi @bnestere , sorry my schedule has been filled with internal work recently. 

Given that the term "slave" has been brought up to be inappropriate large scale - https://jira.mariadb.org/browse/MDEV-18264 and https://jira.mariadb.org/browse/MDEV-18777, we just want to rename them to some other name (e.g., "replica") if we need to introduce them as system variables. Or we can revise and wrap up the three `binlog` related variables in this pull request and open up another pull request later on for other variables.

Thanks a lot, @bnestere !

bnestere commented:
No problem, thanks for the update!

Good question. As you suggest, let's keep the scope of this PR limited to the binlog variables, and then we will create a new ticket for next release to add the rest of the variables. Specifically, we want the new system variables to use the `primary` and `replica` names; and then create aliases for the command-line options to be consistent.

DarrenZhang01 commented:
Hi @bnestere , thanks for the reply! 

I updated the implementation following your insights. Specifically, I added the `Sys_var_binlog_filter`. However, when I started the server with `--binlog-do-db="database1"` option, I see that the according switch-case logic on `OPT_BINLOG_DO_DB` in `mysqld.cc` is called, but when I logged into the server and tried `show variables like "bingo_do_db"`, the variable is there but the value is empty, so I guess the value was not successfully stored in the system variable in the new implementation. Maybe there is something I missed. Would you mind helping check the current logic @bnestere ? (Forgive me some extra printing statement for debugging purposes, and I will clean them up in the final version)

Thanks a lot!

bnestere commented:
Howdy @DarrenZhang01!

On your next commit, can you reword the commit message with the following notes:

1. Update the JIRA # to MDEV-30188 (we created a subtask just for these binlog variables).
2. Update the `note` aspect to be consistent with the code. (Also just as a reminder in your current patch, there is left-over code from your previous version that needs to be removed).

bnestere commented on sql/sys_vars.inl:
```diff
@@ -795,6 +795,51 @@ protected:
+public:
+  Sys_var_binlog_filter(const char *name, int getopt_id, const char *comment,
+                     privilege_t access_global)
+    : sys_var(&all_sys_vars, name, comment, sys_var::GLOBAL, 0, NO_GETOPT,
```
Just a couple other notes, the `flags_arg` will need to be `sys_var::READONLY+sys_var::GLOBAL`. Then below, we don't need `GET_ASK_ADDR`, just `GET_STR` should suffice.

bnestere commented on sql/sys_vars.inl:
```diff
@@ -795,6 +795,51 @@ protected:
+
+  bool do_check(THD *thd, set_var *var) override
+  {
+    return Sys_var_charptr::do_string_check(thd, var, charset(thd));
```
As the variable is read-only, we don't need `do_check()` either, that can also just be (I missed it in my mention initially, sorry)
```
  {
    DBUG_ASSERT(FALSE);
    return true;
  }
```

DarrenZhang01 commented:
Hi @bnestere, thanks for the comments, and I will carefully go through them one by one! Sorry for the delay, and I now continue working on this. 

I added the `global_value_ptr` function override under `Sys_var_binlog_filter` following your insights. Now I realize one issue: In `global_value_ptr`, the object initialized by the `MasterInfo` class in `global_value_ptr` returns the `rpl_filter`, and the `rpl_filter` will be responsible for getting back the information for the variables requested. 

However, if we use the same `rpl_filter` as that in `Sys_var_rpl_filter::global_value_ptr`, the information returned is actually about the replication variables (`replicate_do_db` and `replicate_ignore_db`) rather than binlog variables (`bingo_do_db` and `bingo_ignore_db`). 

In order to resolve this, it *might* be needed to add a new attribute called `binlog_filter` in the `MasterInfo`  class, which can make things a bit more complicated. What are your insights? Correct me if I am wrong. Thanks very much!

DarrenZhang01 commented:
Hi @bnestere, thanks for the comments, and I will carefully go through them one by one! Sorry for the delay, and I now continue working on this.

I added the `global_value_ptr` function override under `Sys_var_binlog_filter` following your insights. Now I realize one issue: In `global_value_ptr`, the object initialized by the `Master_info` class in `global_value_ptr` returns the `rpl_filter`, and the `rpl_filter` will be responsible for getting back the information for the variables requested.
However, if we use the same `rpl_filter` as that in `Sys_var_rpl_filter::global_value_ptr`, the information returned is actually about the replication variables (`replicate_do_db` and `replicate_ignore_db`) rather than binlog variables (`binlog_do_db` and `binlog_ignore_db`).
In order to resolve this, it might potentially be needed to add a new attribute called `binlog_filter` in the `Master_info` class, which can make things a bit more complicated. What are your insights? Correct me if I am wrong. Thanks very much!

bnestere commented:
Hi @DarrenZhang01, great question! There is already a global `binlog_filter` established (see `rpl_filter.h`) which you should use. For explanation as to why we don't want to extend `master_info`, each primary server that a replica connects to will have its own `master_info` (for multi-source replication). The `rpl_filter` instance in `master_info` is specific to the events which come from only that primary server. Conversely, the `binlog_filter` is applied at the storage engine level, such that it simply not write filtered-out events into the replica's binary log (but it may still execute them). To connect the two concepts together, the rules set by a server's global `binlog_filter` would apply to all events which are allowed by each instance of `master_info`'s `rpl_filter` (as well as any directly executed transactions).

DarrenZhang01 commented:
Hi @bnestere , thanks for the insights! I have finished updating the code. The build process and the tests were all successful on my end. I also started a server testing it. But it is weird that several of the above build processes failed and several of the above build processes succeeded.

bnestere commented on mysql-test/suite/sys_vars/t/binlog_do_db.test:
```diff
@@ -0,0 +1,15 @@
+# Restart the server with "--binlog_do_db='database1' --binlog_do_db='database2'" option
+--let $restart_parameters = "--binlog_do_db='database1' --binlog_do_db='database2'"
+--source include/restart_mysqld.inc
+SELECT @@binlog_do_db AS EXPECT_database1_database2;
```
In each of your new test files, can you add a new line at the end of the file?

bnestere commented on sql/sys_vars.cc:
```diff
@@ -5565,6 +5585,35 @@ Sys_var_rpl_filter::global_value_ptr(THD *thd,
+  String tmp(buf, sizeof(buf), &my_charset_bin);
+  uchar *ret;
+
+  mysql_mutex_unlock(&LOCK_global_system_variables);
```
You don't need to switch locks at all, holding the already attained `LOCK_global_system_variables` is safe.

bnestere commented on mysql-test/suite/sys_vars/t/binlog_row_event_max_size.test:
```diff
@@ -0,0 +1,11 @@
+--echo #
+
+SELECT @@binlog_row_event_max_size AS EXPECT_8192;
+
```
Can you also add a quick test showing it is read only? (Once you update the variable).

bnestere reviewed (commented):
Hi @DarrenZhang01, thanks again for the updated patch! I had a couple last findings that should be addressed. Looks great otherwise.

DarrenZhang01 commented:
Hi @bnestere , thanks very much for the insights! I just submitted a new version based on your feedback!

DarrenZhang01 commented:
Hi @bnestere , thanks very much for the insights! I submitted a new version based on your insights. I will do a git rebase to keep the branch updated with 11.0 upstream.

DarrenZhang01 commented:
Hi @bnestere , thanks very much for the insights! I submitted a new version based on your insights. I will do a `git pull --rebase` to keep the branch updated with 11.0 upstream.

bnestere reviewed (approved):
Looks good, thanks @DarrenZhang01 !

Note that this will go into `11.2`, but the branch hasn't yet been created, so it can't be merged until then. I'll update you when it's ready.

DarrenZhang01 commented:
Thanks @bnestere !

andrelkin commented on sql/sys_vars.cc:
```diff
@@ -5593,6 +5617,12 @@ static Sys_var_rpl_filter Sys_replicate_do_db(
        "actual names of table(s) being updated are checked.",
        PRIV_SET_SYSTEM_GLOBAL_VAR_REPLICATE_DO_DB);
 
+static Sys_var_binlog_filter Sys_binlog_do_db(
```
Just a thought, it'd be great eventually to have the class' method to work in `mysqld_get_one_option` that **actually** "populates" the instance.

andrelkin reviewed (approved):
A nice piece of work! Thank you all, Darren and Brandon especially!

vuvova commented:
@LinuxJedi please fix 32-bit builds that this commit broke. See https://buildbot.mariadb.net/buildbot/builders/kvm-fulltest2/builds/41645/steps/mtr_nm/logs/stdio and/or 8ca18cc91384cbb3cd7ce12feb07ddcaa1a449e3

LinuxJedi commented:
@vuvova sorry I missed that, pulled in your fix with #2674, which now passes on the Foundation 32bit.
